### ISSUE TITLE
Modifying mark set after capturing a continuation

### ISSUE THREAD

**Issue by sabauma** [samth/pycket#182]
Under Racket, the mark set of a captured continuation is the set at the time the continuation was captured. Pycket's current implementation allows mark sets to be modified after the continuation is captured, since the mark sets are attached directly to each continuation and are mutated by `with-continuation-marks`.

Example program:

```racket
#lang racket/base

(let* ([extract
	    (lambda (k) (continuation-mark-set->list
	       (continuation-marks k)
	       'x))]
       [go
	    (lambda (in?)
          (lambda ()
	    	      (let ([k (with-continuation-mark 'x 10
	    			 (begin0
	    			  (with-continuation-mark 'x 11
	    			    (call/cc                            ;; Capture a continuation (k)
	    			     (lambda (k)
	    			       (with-continuation-mark 'x 12    ;; Modify its mark set
	    				 (if in?
	    				     (extract k)
	    				     k)))))
	    			  (+ 2 3)))])
	    		(if in?
	    		    k
	    		    (extract k)))))])
  ((go #t)))

  ;; racket => (11 10)
  ;; pycket => (12 10)
```

The mark set of `k` is modified in Pycket; the `result-expr` of `with-continuation-marks` is invoked in tail position so the current continuation is also `k`.  I suspect a shallow copying of the top continuation frame of `with-continuation-marks` would solve the problem, but is annoying given the number of continuation types in Pycket (though would allow us to also implement composable continuations easily).  Racket probably just copies the entire mark stack along with the call stack to produce a continuation, but that would ruin all of our nice `call/cc` benchmarks.

**Comment by sabauma** [comment]
For reference, this example comes from Racket continuation mark test suite.

**Comment by samth** [comment]
I think the shallow copy sounds right to me.

**Comment by sabauma** [comment]
Fixed by 31737006373202115b528309385136a83092296b
